### ISSUE TITLE
Support HTTP authentication in xhr

### ISSUE THREAD

**Issue by jonashaag** [mbostock/d3#2131]
`d3.xhr` currently doesn't offer a way to supply the `username` and `password` parameters to `XmlHttpRequest.open`. Here is a very simple fix I applied to the generated source code. If you're interested in adding this feature, I can prepare a proper pull request with documentation etc (whatever you require for a pull request to land).

I'm not all too familiar with the consequences of enabling `withCredentials` (which enables authentication for cross-origin requests) by default; an approach that I have taken in this simple preliminary patch. It probably should have to be enabled explicitly, maybe as another parameter in `.authenticate`.

```diff
--- d3.js.orig	2014-11-26 18:01:32.962305596 +0100
+++ d3.js	2014-11-26 18:00:31.557119550 +0100
@@ -1844,7 +1844,7 @@
     };
   }
   function d3_xhr(url, mimeType, response, callback) {
-    var xhr = {}, dispatch = d3.dispatch("beforesend", "progress", "load", "error"), headers = {}, request = new XMLHttpRequest(), responseType = null;
+    var xhr = {}, dispatch = d3.dispatch("beforesend", "progress", "load", "error"), headers = {}, request = new XMLHttpRequest(), responseType = null, username = null, password = null;
     if (d3_window.XDomainRequest && !("withCredentials" in request) && /^(http(s)?:)?\/\//.test(url)) request = new XDomainRequest();
     "onload" in request ? request.onload = request.onerror = respond : request.onreadystatechange = function() {
       request.readyState > 3 && respond();
@@ -1883,6 +1883,12 @@
       mimeType = value == null ? null : value + "";
       return xhr;
     };
+    xhr.authentication = function(username_, password_) {
+      if (!arguments.length) return [username, password];
+      username = username_;
+      password = password_;
+      return xhr;
+    }
     xhr.responseType = function(value) {
       if (!arguments.length) return responseType;
       responseType = value;
@@ -1899,7 +1905,8 @@
     });
     xhr.send = function(method, data, callback) {
       if (arguments.length === 2 && typeof data === "function") callback = data, data = null;
-      request.open(method, url, true);
+      request.withCredentials = true;
+      request.open(method, url, true, username, password);
       if (mimeType != null && !("accept" in headers)) headers["accept"] = mimeType + ",*/*";
       if (request.setRequestHeader) for (var name in headers) request.setRequestHeader(name, headers[name]);
       if (mimeType != null && request.overrideMimeType) request.overrideMimeType(mimeType);
@@ -9212,4 +9219,4 @@
   });
   if (typeof define === "function" && define.amd) define(d3); else if (typeof module === "object" && module.exports) module.exports = d3;
   this.d3 = d3;

```

**Comment by mbostock** [comment]
I’m 90% certain you would not want to enable `withCredentials = true` for requests that you don’t also specify authentication parameters for, but I don’t know off-hand exactly what the impact would be. The gist of this request seems reasonable, though I do wonder if there’s a way to make the d3.xhr API a bit more extensible so that users can make modifications to the underlying XMLHttpRequest object without us need to wrap every part of that API.

**Comment by mbostock** [comment]
This has been fixed in [d3-request 0.4.5](https://github.com/d3/d3-request/releases/tag/v0.4.5). For example:

```js
d3.json("https://httpbin.org/basic-auth/foo/bar")
    .user("foo")
    .password("bar")
    .get(function(error, result) { console.log(result); });
```

This returns:

```json
{
  "authenticated": true,
  "user": "foo"
}
```
